packages/astro/src/server/clerk-middleware.ts

```diff
@@ -0,0 +1,343 @@
+  }
+
+  if (
+    isSatellite &&
+    !isHttpOrHttps(signInUrl) &&
+    isDevelopmentFromSecretKey(opts.secretKey || getSafeEnv(context).sk!)
+  ) {
+    throw new Error(missingSignInUrlInDev);
+  }
+
+  return {
+    proxyUrl,
+    isSatellite,
+    domain,
+  };
+};
+
+// Duplicate from '@clerk/nextjs'
+export const missingDomainAndProxy = `
+Missing domain and proxyUrl. A satellite application needs to specify a domain or a proxyUrl.
+
+1) With middleware
+   e.g. export default clerkMiddleware({domain:'YOUR_DOMAIN',isSatellite:true});
+2) With environment variables e.g.
+   PUBLIC_ASTRO_APP_CLERK_DOMAIN='YOUR_DOMAIN'
+   PUBLIC_ASTRO_APP_CLERK_IS_SATELLITE='true'
+   `;
+
+// Duplicate from '@clerk/nextjs'
+export const missingSignInUrlInDev = `
+Invalid signInUrl. A satellite application requires a signInUrl for development instances.
+Check if signInUrl is missing from your configuration or if it is not an absolute URL
+
+1) With middleware
+   e.g. export default clerkMiddleware({signInUrl:'SOME_URL', isSatellite:true});
+2) With environment variables e.g.
+   PUBLIC_ASTRO_APP_CLERK_SIGN_IN_URL='SOME_URL'
+   PUBLIC_ASTRO_APP_CLERK_IS_SATELLITE='true'`;
+
+function decorateAstroLocal(req: Request, context: APIContext, requestState: RequestState) {
+  const { reason, message, status, token } = requestState;
+  context.locals.authToken = token;
+  context.locals.authStatus = status;
+  context.locals.authMessage = message;
+  context.locals.authReason = reason;
+  context.locals.auth = () => getAuth(req, context.locals);
+  context.locals.currentUser = createCurrentUser(req, context);
+}
+
+async function decorateRequest(
+  locals: APIContext['locals'],
+  res: Response,
+  requestState: RequestState,
+): Promise<Response> {
+  const { reason, message, status, token } = requestState;
+
+  res.headers.set(constants.Headers.AuthToken, token || '');
+  res.headers.set(constants.Headers.AuthStatus, status);
+  res.headers.set(constants.Headers.AuthMessage, message || '');
+  res.headers.set(constants.Headers.AuthReason, reason || '');
+
+  /**
+   * Populate every page with the authObject. This allows for SSR to work properly
+   * without the developer having to wrap manually each page with `ClerkLayout.astro`
+   * ^ ClerkLayout is still needed in order to populate the ssrState store, but it not responsible for passing the data to a page.
+   */
+  if (res.headers.get('content-type') === 'text/html') {
+    const reader = res.body?.getReader();
+    const stream = new ReadableStream({
+      async start(controller) {
+        let { value, done } = await reader!.read();
+        const encoder = new TextEncoder();
+        const decoder = new TextDecoder();
+        while (!done) {
+          const decodedValue = decoder.decode(value);
+
+          /**
+           * Hijack html response to position `__CLERK_ASTRO_DATA__` before the closing `head` html tag
+           */
+          if (decodedValue.includes('</head>')) {
```

Comment:
This could likely be made more efficient if we check against the encoded value instead of decoding & encoding the entire body.

e.g.
```ts
const ENCODED_END_HEAD_MARKER = new TextEncoder().encode('</head>');

...

if (value.includes(ENCODED_END_HEAD_MARKER)) {
  const [p1, p2] = value.split(ENCODED_END_HEAD_MARKER);
  controller.enqueue(p1);
  ...
  controller.enqueue(p2);
} else {
  controller.enqueue(value)
}
```